python/ray/serve/_private/deployment_scheduler.py

```diff
@@ -294,6 +358,98 @@ def _get_node_to_running_replicas(
+           subtracting the resources taken up by running (already
+           scheduled by core) and launching (to-be-scheduled and soft
+           targeting that node) replicas.
+
+        Note that (1) may not be accurate because it uses cached info
+        and there is a delay from fetching data from GCS, and (2) may
+        not be accurate because there can be other actors (not replicas)
+        running in the cluster, and launching replicas may not end up on
+        the node we're targeting. So the information returned from this
+        method is only best effort.
+        """
+
+        available_resources = (
+            self._cluster_node_info_cache.get_available_resources_per_node()
+        )
+        total_resources = self._cluster_node_info_cache.get_total_resources_per_node()
+
+        gcs_info = {node_id: Resources(r) for node_id, r in available_resources.items()}
+
+        # Manually calculate available resources per node by subtracting
+        # launching and running replicas from total resources
+        total_minus_replicas = {
+            node_id: Resources(resources)
+            for node_id, resources in total_resources.items()
+        }
+        for deployment_id, replicas in self._launching_replicas.items():
+            for info in replicas.values():
+                if info.target_node_id:
+                    total_minus_replicas[info.target_node_id] -= self._deployments[
+                        deployment_id
+                    ].required_resources
+        for deployment_id, replicas in self._running_replicas.items():
+            for node_id in replicas.values():
+                total_minus_replicas[node_id] -= self._deployments[
+                    deployment_id
+                ].required_resources
+
+        def custom_min(a: Resources, b: Resources):
+            keys = set(a.keys()) | set(b.keys())
+            res = Resources()
+            for key in keys:
+                res[key] = min(a.get(key), b.get(key))
+            return res
+
+        # Filter by active node ids (alive but not draining)
+        return {
+            node_id: custom_min(
+                gcs_info.get(node_id, Resources()),
+                total_minus_replicas.get(node_id, Resources()),
+            )
+            for node_id in self._cluster_node_info_cache.get_active_node_ids()
+        }
+
+    def _best_fit_node(
+        self, required_resources: Resources, available_resources: Dict[str, Resources]
+    ) -> Optional[str]:
+        """Chooses a node using best fit strategy.
+
+        This strategy picks the node where, if the required resources
+        were to be scheduled on that node, it will leave the smallest
+        remaining space. This minimizes fragmentation of resources.
+        """
+
+        min_remaining_space = Resources(
+            {
+                "CPU": float("inf"),
+                "GPU": float("inf"),
+                "memory": float("inf"),
```

Comment:
We should consider all resources of `required_resources`?

Reply:
Fixed!